Source/sdc/Castro_sdc_util.H

```diff
@@ -0,0 +1,883 @@
+    // stored in U_new. That only really makes sense for the 
+    // case where we have 1 substep. Otherwise, we should just
+    // use the old time solution.
+
+    int nsub = 1;
+    ierr = CONVERGENCE_FAILURE;
+
+    for (int n = 0; n < NUM_STATE; ++n) {
+        U_begin[n] = U_old[n];
+    }
+
+    while (nsub < MAX_NSUB && ierr != NEWTON_SUCCESS) {
+        if (nsub > 1) {
+            for (int n = 0; n < NUM_STATE; ++n) {
+                U_new[n] = U_old[n];
+            }
+        }
+        Real dt_sub = dt_m / nsub;
+
+        for (int isub = 0; isub < nsub; ++isub) {
+            // normalize species 
+            Real sum_rhoX = 0.0_rt;
+            for (int n = 0; n < NumSpec; ++n) {
+                U_begin[UFS + n] = amrex::max(small_x, U_begin[UFS + n]);
+                sum_rhoX += U_begin[UFS + n];
+            }
+            for (int n = 0; n < NumSpec; ++n) {
+                U_begin[UFS + n] *= U_begin[URHO] / sum_rhoX;
+            }
+
+            sdc_newton_solve(dt_sub, U_begin, U_new, C, sdc_iteration, err_out, ierr);
+
+            for (int n = 0; n < NUM_STATE; ++n) {
+                U_begin[n] = U_new[n];
+            }
+        }
+        nsub *= 2;
+    }
+}
+
+AMREX_GPU_HOST_DEVICE AMREX_INLINE
+void 
+sdc_newton_solve(const Real dt_m,
+                 GpuArray<Real, NUM_STATE> const& U_old,
+                 GpuArray<Real, NUM_STATE> & U_new,
+                 GpuArray<Real, NUM_STATE> const& C,
+                 const int sdc_iteration,
+                 Real& err_out,
+                 int& ierr) {
+    // the purpose of this function is to solve the system 
+    // U - dt R(U) = U_old + dt C using a Newton solve.
+    // 
+    // here, U_new should come in as a guess for the new U 
+    // and will be returned with the value that satisfied the 
+    // nonlinear function
+
+    RArray2D Jac;
+
+    // we will do the implicit update of only the terms that 
+    // have reactive sources
+    //
+    // 0 : rho
+    // 1:NumSpec : species 
+    // NumSpec+1 : (rho E) or (rho e)
+
+    GpuArray<Real, NumSpec+2> U_react;
+    GpuArray<Real, NumSpec+2> f_source;
+    GpuArray<Real, AMREX_SPACEDIM> mom_source;
+    GpuArray<Real, NumSpec+2> dU_react;
+    GpuArray<Real, NumSpec+2> f;
+    RArray1D f_rhs;
+
+    const int MAX_ITER = 100;
+
+    ierr = NEWTON_SUCCESS;
+
+    // the tolerance we are solving to may depend on the
+    // iteration
+    Real relax_fac = std::pow(sdc_solver_relax_factor, sdc_order - sdc_iteration - 1);
+    Real tol_dens = sdc_solver_tol_dens * relax_fac;
+    Real tol_spec = sdc_solver_tol_spec * relax_fac;
+    Real tol_ener = sdc_solver_tol_ener * relax_fac;
+
+    // update the momenta for this zone -- they don't react
+    for (int n = 0; n < AMREX_SPACEDIM; ++n) {
+        U_new[UMX+n] = U_old[UMX+n] + dt_m * C[UMX+n];
+    }
+
+    // update the non-reacting species
+    for (int n = 0; n < NumSpec; ++n) {
+        U_new[UFS + n] = U_old[UFS+n] + dt_m * C[UFS + n];
+    }
```

Comment:
this loop is not needed.  In the Fortran, this was for the non-reacting stuff, but that feature was removed.  As written here, it covers all species, which is wrong.